test_calculus.py

```diff
@@ -1,9 +1,67 @@
 """
 test_calculus.py
 """
+
+import math
+import pytest
 import calculus as calc
 
 def test_dummy():
     """ Unit test for dummy function
     just the same test function that Dr Thomay made"""
     assert calc.dummy() == 0
+
+# Test data for various cases
+test_data_tanh = [
+    (math.tanh, -1, 1, 1e-6, 0.0)  # (function, a, b, tol, expected_root)
+]
+
+test_data_1_over_sin = [
+    (lambda x: 1 / math.sin(x) if math.sin(x) != 0 else float('inf'), 3, 4, 1e-6, math.pi)
+]
+
+invalid_interval_data = [
+    (math.tanh, -1, -0.5, 1e-6)  # (function, a, b, tol)
+]
+
+# Test data: singularities near sin(x) = 0
+singularity_data = [
+    (lambda x: 1 / math.sin(x) if math.sin(x) != 0 else float('inf'),
+     3.141592653589793 - 1e-3, 3.141592653589793 + 1e-3, 1e-6),  # near singularity
+    (lambda x: 1 / math.sin(x), 3.1405926535897932, 3.142592653589793,
+     1e-6), # Check behavior near zero
+]
+
+# SciPy Wrapper Tests
+@pytest.mark.parametrize("func, a, b, tol, expected", test_data_tanh + test_data_1_over_sin)
+def test_wrapper(func, a, b, tol, expected):
+    """Test SciPy wrapper implementation."""
+    root = calc.bisection_wrapper(func, a, b, tol)
+    assert math.isclose(root, expected, rel_tol=1e-6), f"Expected {expected}, got {root}"
+
+# Pure Python Implementation Tests
+@pytest.mark.parametrize("func, a, b, tol, expected", test_data_tanh + test_data_1_over_sin)
+def test_pure_python(func, a, b, tol, expected):
+    """Test pure Python implementation."""
+    root = calc.bisection_pure_python(func, a, b, tol)
+    assert math.isclose(root, expected, rel_tol=1e-6), f"Expected {expected}, got {root}"
+
+# Invalid Interval Tests
+@pytest.mark.parametrize("func, a, b, tol", invalid_interval_data)
+def test_invalid_interval(func, a, b, tol):
+    """Test for invalid intervals where func(a) and func(b) do not have opposite signs."""
+    with pytest.raises(ValueError):
+        calc.bisection_wrapper(func, a, b, tol)
+
+    with pytest.raises(ValueError):
+        calc.bisection_pure_python(func, a, b, tol)
+
+# Singularity Tests
+@pytest.mark.parametrize("func, a, b, tol", singularity_data)
+def test_singularities(func, a, b, tol):
```

Comment:
test functions should have a name liked to the original functions since this is a shared file